src/atldld/sync.py

```diff
@@ -260,59 +227,190 @@ def download_parallel_dataset(
+        r_images = rma_all(parameters_images)
+
+        # Extract relevant information
+        def extract_template(data, name):
+            template = np.array(AFFINE_TEMPLATES[name])
+            key = "alignment2d" if name in {"tsv", "tvs"} else "alignment3d"
+            return np.vectorize(lambda x: data[key][x])(template)
+
+        metadata["dataset"] = {
+            "id": r_dataset["id"],
+            "affine_tvr": extract_template(r_dataset, "tvr"),
+            "affine_trv": extract_template(r_dataset, "trv"),
+            "plane_of_section_id": r_dataset["plane_of_section_id"],
+            "section_thickness": r_dataset["section_thickness"],
+        }
+
+        images = []
+        for r_image in r_images:
+            images.append(
+                {
+                    "id": r_image["id"],
+                    "affine_tsv": extract_template(r_image, "tsv"),
+                    "affine_tvs": extract_template(r_image, "tvs"),
+                    "section_number": r_image["section_number"],
+                }
+            )
 
-        img = get_image(image_id, downsample=downsample_img)
-
-        if not include_expression:
-            yield image_id, slice_ref_coordinate, img, df
+        images.sort(key=lambda x: -x["section_number"])  # type: ignore
+        metadata["images"] = images  # type: ignore[assignment]
+        self.metadata = metadata
+
+    def run(
+        self,
+    ) -> Generator[
+        Tuple[int, float, np.ndarray, Optional[np.ndarray], DisplacementField],
+        None,
+        None,
+    ]:
+        """Download entire dataset.
+
+        For each image in the dataset, this function performs the following steps:
+
+            1. Query the API to get the `p, i, r` coordinates of the `detection_xy`.
+            2. One of the `p, i, r` will become the `slice_coordinate`. For
+               coronal datasets it is the `p` and for sagittal ones it is the `r`.
+               In other words we assume that the slice is parallel to
+               one of the axes.
+            3. Use `get_parallel_transform` to get a full mapping between the
+               reference space and the image.
+            4. Download the image (+ potentially the expression image)
+            5. Yield result (order derived from section numbers - highest first)
+
+        Returns
+        -------
+        res_dict : generator
+            Generator yielding consecutive four tuples of
+            (image_id, constant_ref_coordinate, img, df).
+            The `constant_ref_coordinate` is the dimension in the given axis in microns.
+            The `img` is the raw gene expression image with dtype `uint8`.
+            The `df` is the displacement field.
+            Note that the sorting. If `include_expression=True` then last returned image
+            is the processed expression image.
+            That is the generator yield (image_id, p, img, df, img_expr).
+        """
+        if self.metadata is None:
```

Comment:
This does not seem to be consistent with the current constructor which initialized `self.metadata = {}`

Reply:
Also it seems like there is no unit test in case the metadata is empty

Reply:
Done in feb3b63 (first comment).

Reply:
Done in 632f947 (second comment).